library/core/src/array/mod.rs

```diff
@@ -564,15 +525,100 @@ impl<T, const N: usize> [T; N] {
     /// ```
     #[unstable(feature = "array_methods", issue = "76118")]
     pub fn each_mut(&mut self) -> [&mut T; N] {
-        // Unlike in `map`, we don't need a guard here, as dropping a reference
-        // is a noop.
-        let mut out = MaybeUninit::uninit_array::<N>();
-        for (src, dst) in self.iter_mut().zip(&mut out) {
-            dst.write(src);
+        // SAFETY: we know for certain that this iterator will yield exactly `N`
+        // items.
+        unsafe { collect_into_array_unchecked(&mut self.iter_mut()) }
+    }
+}
+
+/// Pulls `N` items from `iter` and returns them as an array. If the iterator
+/// yields fewer than `N` items, this function exhibits undefined behavior.
+///
+/// See [`collect_into_array`] for more information.
+///
+///
+/// # Safety
+///
+/// It is up to the caller to guarantee that `iter` yields at least `N` items.
+/// Violating this condition causes undefined behavior.
+unsafe fn collect_into_array_unchecked<I, const N: usize>(iter: &mut I) -> [I::Item; N]
+where
+    I: Iterator,
+{
```

Comment:
Two ideas for this method:
- Could it require `I: TrustedLen`?  If this was exposed publicly that would be a bad idea, but for an internal implementation detail (where we're allowed to use that trait and can easily remove it if it proves a bad idea) I think it'd be interesting.  A scan over all the callers says it'd be fine (array::IntoIter, Map, and Zip all keep it), and it might be a good sanity check against passing something like a `Filter` to this accidentally, where we probably can't be confident in the correctness.
- Regardless of the previous, I suggest this have a `debug_assert!(N <= iter.size_hint().0);` at the beginning.  It'll only run on the debug builder, but that'll still be a nice error to get if I make a future PR getting this wrong at some point.

Reply:
> Could it require I: TrustedLen?

The advantage of that would also be that we can avoid dropping elements when the source is too short. I.e. the implementation would be guaranteed to be non-lossy.

But as you say, then it can't be used for whatever generalized public interface comes out of #81615 & Co.

Reply:
I added `TrustedLen`, the debug asserts, and a comment about it now.